sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateProjection.scala

```diff
@@ -84,9 +84,15 @@ object GenerateProjection extends CodeGenerator[Seq[Expression], Projection] {
         val elementName = newTermName(s"c$i")
         val iLit = ru.Literal(Constant(i))
 
-        q"if(i == $ordinal) { if(isNullAt($i)) return null else return $elementName }"
+        q"if(i == $ordinal) { return $elementName }"
       }
-      q"override def apply(i: Int): Any = { ..$cases; $accessorFailure }"
+      q"""
+        override def apply(i: Int): Any = {
+          if (i < 0 || i >= this.length) $accessorFailure
```

Comment:
It is used to check if the parameter `i` fails in legal range. `if(i == $ordinal)` is used to choose the corresponding element to return. Checking the legal range and giving `accessorFailure` in advance can avoid further running of `$cases`. (when row size grows, that will be many `if` checks.)